src/rules/noUnnecessaryInitializerRule.ts

```diff
@@ -40,62 +40,58 @@ export class Rule extends Lint.Rules.AbstractRule {
         "Also, the type declaration does not need to include '| undefined'.";
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithWalker(new Walker(sourceFile, this.getOptions()));
+        return this.applyWithFunction(sourceFile, walk);
     }
 }
 
-class Walker extends Lint.RuleWalker {
-    public visitVariableDeclaration(node: ts.VariableDeclaration) {
-        if (utils.isBindingPattern(node.name)) {
-            for (const elem of node.name.elements) {
-                if (elem.kind === ts.SyntaxKind.BindingElement) {
-                    this.checkInitializer(elem);
+function walk(ctx: Lint.WalkContext<void>) {
+    ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        switch (node.kind) {
+            case ts.SyntaxKind.VariableDeclaration: {
+                const { name } = node as ts.VariableDeclaration;
+                if (utils.isBindingPattern(name)) {
+                    for (const elem of name.elements) {
```

Comment:
this must be recursive.
I'd suggest to remove the special handling of `BindingPattern` here and add a case for `BindingElement` to the switch. That way you also handle destructuring in `ParameterDeclaration`